File: 33.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        ans = -1
        l, r = 0, len(nums) - 1
        while l < r:
            m = (r - l) // 2 + l
            if nums[m] > nums[r]:
                l = m + 1
            else:
                r = m
        pivot = l
        l = 0
        r = len(nums) - 1
        # if the minimum elt in the left portion is still greater than target,
        # look right
        if nums[pivot] <= target and target <= nums[r]:
            l = pivot
        # otherwise, look in the left portion
        else:
            r = pivot
        while l <= r:
            m = (r - l) // 2 + l
            if nums[m] > target:
                r = m - 1
            elif nums[m] < target:
                l = m + 1
            else:
                return m
        return ans


Solution().search([1, 3], 3)

File: 33.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
public:
  int search(vector<int> &nums, int target) {
    int l = 0, r = nums.size() - 1;
    while (l < r) {
      int m = (r - l) / 2 + l;
      if (nums[m] > nums[r]) {
        l = m + 1;
      } else {
        r = m;
      }
    }
    int pivot = l;
    l = 0, r = nums.size() - 1;
    if (nums[pivot] <= target && target <= nums[r]) {
      l = pivot;
    } else {
      r = pivot;
    }
    while (l <= r) {
      int m = (r - l) / 2 + l;
      if (nums[m] > target) {
        r = m - 1;
      } else if (nums[m] < target) {
        l = m + 1;
      } else {
        return m;
      }
    }
    return -1;
  }
};
int main() {
  vector<int> t_in = {1, 3};
  auto res = Solution().search(t_in, 3);
  cout << res << endl;
}